src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobExecutionService.java

```diff
@@ -453,32 +463,52 @@ private BuildResult constructBuildResult(List<BuildResult.LocalCITestJobDTO> fai
     }
 
     private Path cloneRepository(VcsRepositoryUri repositoryUri, @Nullable String commitHash, boolean checkout, String buildJobId) {
+        int attempt = 0;
+        Repository repository = null;
+
+        while (attempt < MAX_CLONE_RETRIES) {
+            try {
+                attempt++;
+                String commitHashForPath = commitHash != null ? commitHash : generateCommitHash();
+                // Clone the assignment repository into a temporary directory
+                repository = buildJobGitService.cloneRepository(repositoryUri, Path.of(CHECKED_OUT_REPOS_TEMP_DIR, commitHashForPath, repositoryUri.folderNameForRepositoryUri()));
+
+                break;
+            }
+            catch (GitAPIException | IOException | URISyntaxException e) {
+                if (attempt >= MAX_CLONE_RETRIES) {
+                    String msg = "Error while cloning repository " + repositoryUri.repositorySlug() + " with uri " + repositoryUri + " after " + MAX_CLONE_RETRIES + " attempts";
+                    buildLogsMap.appendBuildLogEntry(buildJobId, msg);
+                    throw new LocalCIException(msg, e);
+                }
+                buildLogsMap.appendBuildLogEntry(buildJobId,
+                        "Attempt " + attempt + " to clone repository " + repositoryUri.repositorySlug() + " failed due to " + e.getMessage() + ". Retrying...");
+            }
+        }
+
         try {
-            // Clone the assignment repository into a temporary directory
-            // TODO: use a random value if commitHash is null
-            Repository repository = buildJobGitService.cloneRepository(repositoryUri, Path.of(CHECKED_OUT_REPOS_TEMP_DIR, commitHash, repositoryUri.folderNameForRepositoryUri()));
             if (checkout && commitHash != null) {
                 // Checkout the commit hash
                 buildJobGitService.checkoutRepositoryAtCommit(repository, commitHash);
             }
+
             // if repository is not closed, it causes weird IO issues when trying to delete the repository later on
             // java.io.IOException: Unable to delete file: ...\.git\objects\pack\...
             repository.closeBeforeDelete();
             return repository.getLocalPath();
         }
-        catch (GitAPIException | IOException | URISyntaxException e) {
-            String msg = "Error while cloning repository " + repositoryUri.repositorySlug() + " with uri " + repositoryUri;
+        catch (GitException e) {
+            String msg = "Error while checking out commit " + commitHash + " in repository " + repositoryUri.repositorySlug();
             buildLogsMap.appendBuildLogEntry(buildJobId, msg);
             throw new LocalCIException(msg, e);
         }
     }
 
-    private void deleteCloneRepo(VcsRepositoryUri repositoryUri, @Nullable String commitHash, String buildJobId) {
+    private void deleteCloneRepo(VcsRepositoryUri repositoryUri, @Nullable String commitHash, String buildJobId, Path repoPath) {
```

Comment:
nitpick: I guess when using `repositoryUrl`, we should also use `repositoryPath`